text/0092-contract-source-registry.md

```diff
@@ -0,0 +1,159 @@
+7. [Working demo showing the client for publishing and verifying sources](https://ton-defi-org.github.io/ton-src-webclient/EQDerEPTIh0O8lBdjWc6aLaJs5HYqlfBN2Ruj1lJQH_6vcaZ)
+
+# Specification
+
+Due to the large number of moving parts, the specification at this stage does not go into the full details of exact interfaces or TL-Bs. These parts will be added for community discussion in a later stage.
+
+## General
+
+* Contract sources are indexed by code hash and not by contract address. This means that several deployed instances of the same smart contract code will only need to be registered once.
+
+* If a smart contract supports code updates, once its code is changed, the new code hash will no longer be found in the registry until a source code matching the new code hash is uploaded.
+
+* IPFS is used as a **temporary** form of storage only until TON Storage is released. All participants commit to performing a full migration of all verified sources to TON Storage when possible.
+
+* A single participating **verifier** runs a backend that exposes HTTP API that receives a list of source files, compiles them and returns a signed proof over the resulting code hash. The public key for this verifier is registered in the verifier registry contract.
+
+* To encourage redundancy of verifiers, each verifier is encouraged to run multiple backend instances, each with its own public key. Verification then requires a quorum of backends to work and not all of them, for example 3 out of 5. This is optional and a verifier can choose to run a single backend with a single public key too. This redundancy allows for easy participation of decentralized oracle networks like Orbs Network in the protocol (networks that have dozens of independent validators) as a single verifier.
+
+## sources.json
+
+JSON file provided by a specific **verifier** for a specific **contract code hash** containing the URLs of source-code files and verification attestations. Fields of this file include:
+
+* `codeHash` - SHA256 of the code cell CellRepr (https://ton-blockchain.github.io/docs/tvm.pdf 3.1.5), e.g. E/XXoxbG124QU+iKxZtd5loHKjiEUTcdxcW+y7oT9Q4=
+* `sources` - Array of sources files, each with the following fields:
+    * `url` - Human-readable source-code URL on IPFS and later TON Storage, e.g. [ipfs://QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T](https://tonsource.infura-ipfs.io/ipfs/QmWQE1HhYuWieFcocfZZvK4gvJqgswNpj3zUCvdPNCC12T)
+    * `originalFilename` - e.g. "wallet-v3.fc" or "imports/stdlib.fc"
+* `compilerType` - Compiler used to compile the verified sources, e.g. "func"
+* `compilerVersion` - Version of compiler, e.g. "0.2.0"
+* `commandLine` - Command-line used to compile sources, e.g. "func -o output.fif -SPA wallet-v3.fc stdlib.fc"
+* `verificationTime` - Unix timestamp of when verification was made
+
```

Comment:
Why do we need to keep things so complicated? 

In my opinion, it is much more reasonable to use 1 link to the project in ipfs (onchain), and also to store hash of on-chain code giving it in get method together with verifier_id to make such contracts easier to index. Making a link to each source file is very difficult.

It is also necessary that this project contains a configuration file by the proposed standard [TEP: FunC project configuration file](https://github.com/ton-blockchain/TEPs/pull/83). I believe that this standard can unify all CLI tools and will work much more reliably for building projects.